manuals/ewfverify.1
.Dd January 19, 2014
.Dt ewfverify
.Os libewf
.Sh NAME
.Nm ewfverify
.Nd verifies media data stored in EWF files
.Sh SYNOPSIS
.Nm ewfverify
.Op Fl A Ar codepage
.Op Fl d Ar digest_type
.Op Fl f Ar format
.Op Fl l Ar log_filename
.Op Fl p Ar process_buffer_size
.Op Fl hqvVwx
.Ar ewf_files
.Sh DESCRIPTION
.Nm ewfverify
is a utility to verify media data stored in EWF files.
.Pp
.Nm ewfverify
is part of the
.Nm libewf
package.
.Nm libewf
is a library to access the Expert Witness Compression Format (EWF).
.Pp
.Ar ewf_files
the first or the entire set of EWF segment files
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A Ar codepage
the codepage of header section, options: ascii (default), windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252, windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258
.It Fl d Ar digest_type
calculate additional digest (hash) types besides md5, options: sha1, sha256
.It Fl f Ar format
specify the input format, options: raw (default), files (restricted to logical volume files)
.It Fl h
shows this help
.It Fl l Ar log_filename
logs verification errors and the digest (hash) to the log filename
.It Fl p Ar process_buffer_size
the process buffer size (default is the chunk size)
.It Fl q
quiet shows minimal status information
.It Fl v
verbose output to stderr
.It Fl V
print version
.It Fl w
zero sectors on checksum error (mimic EnCase like behavior)
.It Fl x
use the chunk data instead of the buffered read and write functions.
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Nm ewfverify
can either verify the raw storage media data or logical file data stored in EWF file(s).
.Ss To verify the raw storage media data:
.Bd -literal
# ewfverify floppy.E01
ewfverify 20110801

Verify started at: Tue Jan 11 19:21:51 2011

This could take a while.

Status: at 2%.
        verified 32 kB (32768 bytes) of total 1.4 MiB (1474560 bytes).

.Dl ...

Status: at 100%.
        verified 1.4 MiB (1474560 bytes) of total 1.4 MiB (1474560 bytes).
        completion in 1 second(s) with 1 MiB/s (1474560 bytes/second).

Verify completed at: Tue Jan 11 19:21:52 2011

Read: 1.4 MiB (1474560 bytes) in 1 second(s) with 1 MiB/s (1474560 bytes/second).

MD5 hash stored in file:		ae1ce8f5ac079d3ee93f97fe3792bda3
MD5 hash calculated over data:		ae1ce8f5ac079d3ee93f97fe3792bda3

ewfverify: SUCCESS
.Ed
.Ss To verify logical file data:
.Bd -literal
# ewfverify -f files logical.E01
ewfverify 20110111

Verify started at: Tue Jan 11 19:21:51 2011

Single file: file1.txt
MD5 hash stored in file:		6fef2a64afbf070a264366183698bb5e
MD5 hash calculated over data:		6fef2a64afbf070a264366183698bb5e

.Dl ...

Verify completed at: Tue Jan 11 19:21:52 2011

ewfverify: SUCCESS
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \-v is enabled. Verbose and debug output are only printed when enabled at compilation.
.Sh BUGS
Please report bugs of any kind on the project issue tracker: https://github.com/libyal/libewf-legacy/issues
.Sh AUTHOR
.Pp
These man pages were written by Kees Mastwijk.
.Pp
Alterations for distribution have been made by Joachim Metz.
.Sh COPYRIGHT
.Pp
Copyright 2006-2014, Joachim Metz <joachim.metz@gmail.com>.
.Pp
This is free software; see the source for copying conditions. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.Sh SEE ALSO
.Xr ewfacquire 1 ,
.Xr ewfacquirestream 1 ,
.Xr ewfexport 1 ,
.Xr ewfinfo 1 ,
.Xr ewfmount 1 ,
.Xr ewfrecover 1


ChangeLog
TODO
* sync with experimental version
* move mfdata to fdata - not possible
  - chunk_table_list - needs mfdata uses grouping
  - file_list - not supported by fdata

20140801
* bug fix in Python-bindings

20140608
* bug fix for utf16 header functions

20140602
* bug fix in ewfmount regarding logical files date and time values

20140525
* updated python.m4

20140427
* fixes to build static library with mingw and cygwin

20140406
* bug fixes in m4 files

20140228
* removed #error restriction in dependency include header files
* make pyewf_handle_open more strict to catch non-string objects
  without the check the code will segfault on non-string objects

20140227
* bug fixes in empty block compression

20140226
* bug fix in libewf_read_io_handle_read_chunk_data error tolerance
  code path

20140216
* bug fix in recent process status changes

20140211
* integrating latest update for multi threaded ewfacquire changes

20140201
* changed behavior of empty-block check
* worked on integrating multi threaded ewfacquire changes

20140119
* 2014 update
* updated dependencies
* added libcdatetime
* removed borlandc files

20131230
* small updates

20131222
* moved low-level function support from compile time to run time
* worked on sync with experimental version

20131210
* updated dependencies
* worked on Python bindings
* added libcthreads
* fix in DFXML output for size values
* worked on ewfmount

20130416
* bug fix in libcfile.m4 for building on MingW and Cygwin
* fix in libsmdev for MinGW build

20130402
* changes to zlib.m4 for adler32 detection
* moved README.mingw and README.static to wiki

20130331
* worked on libcdata integration
* updated msvscpp files
* updated codegear files
* sync with experimental version

20130330
* updated dependencies
* updates for libsmdev
* added error tolerance for Logicube image with missing checksum in data section
* code clean up
* worked on libcdata integration

20130302
* updated types.h
* fixed maximum number of segments

20130205
* changes and fixes in debug output

20130128
* worked on sync with experimental version
* docstring changes in pyewf
* fix for corruption scenario

20130126
* fixes in pyewf examples

20130120
* updated msvscpp files
* updated codegear files
* updated pyewf
* worked on sync with experimental version
* replace libmfcache by new libfcache

20130119
* updated configure files
* updated dpkg files
* updated rpm spec file
* updated pyewf - fixes multiple issues
* updated dependencies
* worked on sync with experimental version
* added pyewf/setup.py with thanks to Michael Cohen

20130105
* bug fix for 31th day of the month issue

20121209
* removed memwatch and gettext

20120819
* bug fix for not writing SHA-1 hash

20120603
* bug fix for MinGW in liberror
* updates for zlib-1.2.7

20120504
* bug fix for setting media size of L01 file

20120421
* small changes to dpkg files

20120416
* added fuse destroy function to ewfmount
* bug fix in libewf_utf8_string_day_of_month_copy_to_time_elements
* changes to debug output

20120414
* ewfverify: fixed printing the same segment filename multiple times for CRC errors
* worked on pyewf
* bug fix in libewf_utf16_string_month_copy_from_time_elements
* updated libuna: fix in codepage 932

20120304
* small changes to debug error output
* small changes for error tolerability
* updated libfvalue

20120225
* promoted to stable

20120213
* small changes to ewftools
* pyewf: added -avoid-version to Makefile.am
* updated macosx package maker files

20120212
* update for zlib-1.2.6
* added API functions for file entry data offset, data size and
  duplicate data offset
* added file entry flags and media data value functions to ewf.net
* updated libnotify, libodraw, libfvalue, libsmraw

20120209
* small changes

20120208
* added dpkg files

20120122
* bug fix in write finalize

20120118
* updated gettext files
* updated libbfio
* preparations for beta testing

20120111
* updated spec file

20120102
* 2012 update

20111231
* updated libhmac, libsmraw, libsmdev

20111228
* updated libcstring, libuna, libbfio
* updated msvscpp, builder5 and codegear files

20111226
* updated libbfio

20111224
* updated libsystem, libuna, libbfio, libhmac, libfvalue, libodraw

20111221
* small fixes in ewftools

20111208
* worked on dealing with missing segment files

20111207
* worked on dealing with missing segment files

20111205
* worked on ewfrecover

20111204
* updated msvscpp files
* updated builder5 files
* updated codegear files

20111203
* worked on ewfrecover
* code clean in ewftools
* updated libcstring, libsystem
* added libmfcache

20111121
* fixed bug introduced by error tollerance changes

20111120
* updated libfvalue, libmfdata
* worked on error tollerance

20111117
* updated libmfdata

20111115
* code clean up
* updated libmfdata

20111114
* code clean up
* updated array type, list type, tree type
* updated libcstring, libfvalue, libhmac

20111105
* worked on libewf_file_get_file_entry_by_utf8_path and
  libewf_file_get_file_entry_by_utf16_path
* changes to ewfmount to deal with file names containing
  the path separator character /

20111103
* worked on libewf_file_get_file_entry_by_utf8_path and
  libewf_file_get_file_entry_by_utf16_path
* code clean up changed name of file entry item flags and moved it to
  definitions

20111101
* small changes

20111031
* updated README.macosx
* added setvbuf
* updated libcstring, libsystem, libuna, libbfio, libfvalue, libhmac
* added windows codepage 932 and 936 support

20111030
* updated configure.ac and m4 files
* updated libhmac, libodraw, libsmraw, libfvalue, libsmdev
* updated tree type
* changes for libfvalue

20111023
* worked on ewf.net
* bug fixes in ewfmount

20111020
* updated README.static

20111017
* worked on ewf.net file entry support
* changes to m4/libuuid.m4

20111016
* added libewf_file_entry_get_offset
* worked on pyewf file entry support
* changed file entry date and time values to 32-bit values
* worked on ewfmount LEV support

ewftools/ewfverify.c
/*
 * Verifies the integrity of the media data within the EWF file
 *
 * Copyright (c) 2006-2014, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_SYS_RESOURCE_H )
#include <sys/resource.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_GLOB_H )
#include <glob.h>
#endif

#include "byte_size_string.h"
#include "digest_hash.h"
#include "ewfcommon.h"
#include "ewftools_getopt.h"
#include "ewftools_glob.h"
#include "ewftools_libcerror.h"
#include "ewftools_libclocale.h"
#include "ewftools_libcnotify.h"
#include "ewftools_libewf.h"
#include "ewftools_output.h"
#include "ewftools_signal.h"
#include "ewftools_unused.h"
#include "log_handle.h"
#include "verification_handle.h"

verification_handle_t *ewfverify_verification_handle = NULL;
int ewfverify_abort                                  = 0;

/* Prints the executable usage information to the stream
 */
void usage_fprint(
      FILE *stream )
{
	if( stream == NULL )
	{
		return;
	}
	fprintf( stream, "Use ewfverify to verify data stored in the EWF format (Expert Witness\n"
	                 "Compression Format).\n\n" );

	fprintf( stream, "Usage: ewfverify [ -A codepage ] [ -d digest_type ] [ -f format ]\n"
	                 "                 [ -l log_filename ] [ -p process_buffer_size ]\n"
	                 "                 [ -hqvVwx ] ewf_files\n\n" );

	fprintf( stream, "\tewf_files: the first or the entire set of EWF segment files\n\n" );

	fprintf( stream, "\t-A:        codepage of header section, options: ascii (default),\n"
	                 "\t           windows-874, windows-932, windows-936, windows-949,\n"
	                 "\t           windows-950, windows-1250, windows-1251, windows-1252,\n"
	                 "\t           windows-1253, windows-1254, windows-1255, windows-1256,\n"
	                 "\t           windows-1257 or windows-1258\n" );
	fprintf( stream, "\t-d:        calculate additional digest (hash) types besides md5,\n"
	                 "\t           options: sha1, sha256\n" );
	fprintf( stream, "\t-f:        specify the input format, options: raw (default),\n"
	                 "\t           files (restricted to logical volume files)\n" );
	fprintf( stream, "\t-h:        shows this help\n" );
	fprintf( stream, "\t-l:        logs verification errors and the digest (hash) to the\n"
	                 "\t           log_filename\n" );
	fprintf( stream, "\t-p:        specify the process buffer size (default is the chunk size)\n" );
	fprintf( stream, "\t-q:        quiet shows minimal status information\n" );
	fprintf( stream, "\t-v:        verbose output to stderr\n" );
	fprintf( stream, "\t-V:        print version\n" );
	fprintf( stream, "\t-w:        zero sectors on checksum error (mimic EnCase like behavior)\n" );
	fprintf( stream, "\t-x:        use the chunk data instead of the buffered read and write\n"
	                 "\t           functions.\n" );
}

/* Signal handler for ewfverify
 */
void ewfverify_signal_handler(
      ewftools_signal_t signal EWFTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function   = "ewfverify_signal_handler";

	EWFTOOLS_UNREFERENCED_PARAMETER( signal )

	ewfverify_abort = 1;

	if( ewfverify_verification_handle != NULL )
	{
		if( verification_handle_signal_abort(
		     ewfverify_verification_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal verification handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
#if defined( HAVE_GETRLIMIT )
	struct rlimit limit_data;
#endif
	libcerror_error_t *error                                      = NULL;

	system_character_t * const *argv_filenames         = NULL;

#if !defined( HAVE_GLOB_H )
	ewftools_glob_t *glob                                         = NULL;
#endif

	system_character_t *log_filename                   = NULL;
	system_character_t *program                        = _SYSTEM_STRING( "ewfverify" );
	system_character_t *option_additional_digest_types = NULL;
	system_character_t *option_format                  = NULL;
	system_character_t *option_header_codepage         = NULL;
	system_character_t *option_process_buffer_size     = NULL;

	log_handle_t *log_handle                                      = NULL;

	system_integer_t option                            = 0;
	uint8_t calculate_md5                                         = 1;
	uint8_t print_status_information                              = 1;
	uint8_t use_chunk_data_functions                              = 0;
	uint8_t verbose                                               = 0;
	uint8_t zero_chunk_on_error                                   = 0;
	int number_of_filenames                                       = 0;
	int result                                                    = 0;

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
             "ewftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( ewftools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		ewftools_output_version_fprint(
		 stderr,
		 program );

		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	ewftools_output_version_fprint(
	 stdout,
	 program );

	while( ( option = ewftools_getopt(
	                   argc,
	                   argv,
	                   _SYSTEM_STRING( "A:d:f:hl:p:qvVwx" ) ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				usage_fprint(
				 stdout );

				goto on_error;

			case (system_integer_t) 'A':
				option_header_codepage = optarg;

				break;

			case (system_integer_t) 'd':
				option_additional_digest_types = optarg;

				break;

			case (system_integer_t) 'f':
				option_format = optarg;

				break;

			case (system_integer_t) 'h':
				usage_fprint(
				 stdout );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 'p':
				option_process_buffer_size = optarg;

				break;

			case (system_integer_t) 'q':
				print_status_information = 0;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				ewftools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'w':
				zero_chunk_on_error = 1;

				break;

			case (system_integer_t) 'x':
				use_chunk_data_functions = 1;

				break;
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing EWF image file(s).\n" );

		usage_fprint(
		 stdout );

		goto on_error;
	}
	libcnotify_verbose_set(
	 verbose );

#if !defined( HAVE_LOCAL_LIBEWF )
	libewf_notify_set_verbose(
	 verbose );
	libewf_notify_set_stream(
	 stderr,
	 NULL );
#endif

	if( verification_handle_initialize(
	     &ewfverify_verification_handle,
	     calculate_md5,
	     use_chunk_data_functions,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to create verification handle.\n" );

		goto on_error;
	}
	if( option_header_codepage != NULL )
	{
		result = verification_handle_set_header_codepage(
			  ewfverify_verification_handle,
		          option_header_codepage,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set header codepage.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported header codepage defaulting to: ascii.\n" );
		}
	}
	if( option_format != NULL )
	{
		result = verification_handle_set_format(
			  ewfverify_verification_handle,
			  option_format,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set format.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported input format defaulting to: raw.\n" );
		}
	}
	if( option_process_buffer_size != NULL )
	{
		result = verification_handle_set_process_buffer_size(
			  ewfverify_verification_handle,
			  option_process_buffer_size,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set process buffer size.\n" );

			goto on_error;
		}
		else if( ( result == 0 )
		      || ( ewfverify_verification_handle->process_buffer_size > (size_t) SSIZE_MAX ) )
		{
			ewfverify_verification_handle->process_buffer_size = 0;

			fprintf(
			 stderr,
			 "Unsupported process buffer size defaulting to: chunk size.\n" );
		}
	}
	if( option_additional_digest_types != NULL )
	{
		result = verification_handle_set_additional_digest_types(
			  ewfverify_verification_handle,
			  option_additional_digest_types,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set additional digest types.\n" );

			goto on_error;
		}
	}
#if !defined( HAVE_GLOB_H )
	if( ewftools_glob_initialize(
	     &glob,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize glob.\n" );

		goto on_error;
	}
	if( ewftools_glob_resolve(
	     glob,
	     &( argv[ optind ] ),
	     argc - optind,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to resolve glob.\n" );

		goto on_error;
	}
	if( ewftools_glob_get_results(
	     glob,
	     &number_of_filenames,
	     (system_character_t ***) &argv_filenames,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to retrieve glob results.\n" );

		goto on_error;
	}
#else
	argv_filenames      = &( argv[ optind ] );
	number_of_filenames = argc - optind;
#endif

#if defined( HAVE_GETRLIMIT )
	if( getrlimit(
            RLIMIT_NOFILE,
            &limit_data ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to determine limit: number of open file descriptors.\n" );
	}
	if( limit_data.rlim_max > (rlim_t) INT_MAX )
	{
		limit_data.rlim_max = (rlim_t) INT_MAX;
	}
	if( limit_data.rlim_max > 0 )
	{
		limit_data.rlim_max /= 2;
	}
	if( verification_handle_set_maximum_number_of_open_handles(
	     ewfverify_verification_handle,
	     (int) limit_data.rlim_max,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set maximum number of open file handles.\n" );

		goto on_error;
	}
#endif
	if( ewftools_signal_attach(
	     ewfverify_signal_handler,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to attach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	result = verification_handle_open_input(
	          ewfverify_verification_handle,
	          argv_filenames,
	          number_of_filenames,
	          &error );

	if( ewfverify_abort != 0 )
	{
		goto on_abort;
	}
	if( result != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open EWF image file(s).\n" );

		goto on_error;
	}
#if !defined( HAVE_GLOB_H )
	if( ewftools_glob_free(
	     &glob,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free glob.\n" );

		goto on_error;
	}
#endif
	if( verification_handle_set_zero_chunk_on_error(
	     ewfverify_verification_handle,
	     zero_chunk_on_error,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set zero on chunk error.\n" );

		goto on_error;
	}
	if( log_filename != NULL )
	{
		if( log_handle_initialize(
		     &log_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to create log handle.\n" );

			goto on_error;
		}
		if( log_handle_open(
		     log_handle,
		     log_filename,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to open log file: %" PRIs_SYSTEM ".\n",
			 log_filename );

			goto on_error;
		}
	}
	if( ewfverify_verification_handle->input_format == VERIFICATION_HANDLE_INPUT_FORMAT_FILES )
	{
		result = verification_handle_verify_single_files(
		          ewfverify_verification_handle,
		          print_status_information,
		          log_handle,
		          &error );

		if( result != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to verify single files.\n" );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	else
	{
		result = verification_handle_verify_input(
		          ewfverify_verification_handle,
		          print_status_information,
		          log_handle,
		          &error );

		if( result != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to verify input.\n" );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	if( log_handle != NULL )
	{
		if( log_handle_close(
		     log_handle,
		     &error ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to close log handle.\n" );

			goto on_error;
		}
		if( log_handle_free(
		     &log_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to free log handle.\n" );

			goto on_error;
		}
	}
on_abort:
	if( ewftools_signal_detach(
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to detach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( verification_handle_close(
	     ewfverify_verification_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close verification handle.\n" );

		goto on_error;

	}
	if( verification_handle_free(
	     &ewfverify_verification_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free verification handle.\n" );

		goto on_error;
	}
	if( ewfverify_abort != 0 )
	{
		fprintf(
		 stdout,
		 "%" PRIs_SYSTEM ": ABORTED\n",
		 program );

		return( EXIT_FAILURE );
	}
	if( result != 1 )
	{
		fprintf(
		 stdout,
		 "%" PRIs_SYSTEM ": FAILURE\n",
		 program );

		return( EXIT_FAILURE );
	}
	fprintf(
	 stdout,
	 "%" PRIs_SYSTEM ": SUCCESS\n",
	 program );

	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( log_handle != NULL )
	{
		log_handle_close(
		 log_handle,
		 NULL );
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	if( ewfverify_verification_handle != NULL )
	{
		verification_handle_close(
		 ewfverify_verification_handle,
		 NULL );
		verification_handle_free(
		 &ewfverify_verification_handle,
		 NULL );
	}
#if !defined( HAVE_GLOB_H )
	if( glob != NULL )
	{
		ewftools_glob_free(
		 &glob,
		 NULL );
	}
#endif
	return( EXIT_FAILURE );
}

